### PR TITLE
port of magic_gen.py to magic_gen.go

### PR DESCRIPTION
**Pull request by Pac23**
diverged from large pr to this. solves #16

### CODE REVIEW

File: crit-go/scripts/magic_gen.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"github.com/thoas/go-funk"
+	"os"
+	"reflect"
+	"strconv"
+	"strings"
+	//"io/ioutil"
```
Comment (adrianreber) [inline crit-go/scripts/magic_gen.go:12]: Please remove all uncommented code from the file, no need to keep that in git.

File: crit-go/scripts/magic_gen.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+)
+
+func main() {
+
+	if len(os.Args) != 3 {
+		//fmt.Println("Usage: magic-gen.py path/to/image.h path/to/magic.json")
```
Comment (adrianreber) [inline crit-go/scripts/magic_gen.go:18]: This looks like a useful message to print out

File: crit-go/scripts/magic_gen.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+		os.Exit(1)
+	}
+	magic_c_header := os.Args[1]
+	magic_json := os.Args[2]
+	infile, err := os.Open(magic_c_header) // opens magic file
```
Comment (adrianreber) [inline crit-go/scripts/magic_gen.go:23]: Please no comments at the end of the line. Move it on its own life above.

File: crit-go/scripts/magic_gen.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	for k, v := range magic {
+		//fmt.Println(reflect.TypeOf(v))
+		if v == "0x0" || v == "1" || k == "0x0" || v == "1" {
+			continue
+		}
+		if strings.Contains(k, "_MAGIC") == true {
+			k = strings.Replace(k, "_MAGIC", "", -1)
+			//fmt.Println(k)
```
Comment (adrianreber) [inline crit-go/scripts/magic_gen.go:72]: Remove that or add a debug mode.

**Comment by adrianreber** [comment]
Please add a Makefile to build your tool and make sure it is also built during a Travis run.

**Comment by adrianreber** [comment]
Also add some information to the commit message about this tool. Why we need it and how to use it.

**Comment by adrianreber** [comment]
You could also run a minimal test during Travis with a minimal input file and compare the output json.

**Comment by adrianreber** [comment]
Please rebase to take your Travis fixes into this PR.

As long as the code is not yet committed to the main repository please no fix up commits. Just fix the existing commit. In this case you need to squash both commits to one commit and force push to your branch.

**Comment by Pac23** [comment]
> Also add some information to the commit message about this tool. Why we need it and how to use it.

Done 

> Please rebase to take your Travis fixes into this PR.
> 
> As long as the code is not yet committed to the main repository please no fix up commits. Just fix the existing commit. In this case you need to squash both commits to one commit and force push to your branch.

Fixed,squashed with parent.

**Comment by adrianreber** [comment]
There are still two commented out lines in the file: fmt.Println() , please remove them

Please edit the commit message to not look like two commits messages. Make one commit message out of it. Stay under 80 characters per line in the commit message (except for error messages or examples). Add an example how you can use your program.

Have a look at https://github.com/checkpoint-restore/go-criu/blob/master/Makefile and add a step to build and lint your program. Add a test run with a minimal input (maybe only two lines) which you can easily compare maybe in the Makefile. Add the built, the lint and the test to .travis.yml.

File: crit-go/scripts/magic_gen.go [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+	"os"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+func main() {
+
+	if len(os.Args) != 3 {
+		fmt.Println("Usage: magic-gen.py path/to/image.h path/to/magic.json")
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:17]: It's no longer `.py` I guess.

Comment (kolyshkin) [reply crit-go/scripts/magic_gen.go:17]: .. and if you want to call it `magic-gen`, put it into `magic-gen` directory.

File: crit-go/scripts/magic_gen.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	scanner := bufio.NewScanner(infile)
+	scanner.Split(bufio.ScanLines)
+
+	var txtlines []string
+
+	for scanner.Scan() {
+		txtlines = append(txtlines, scanner.Text())
+	}
+
+	for _, eachline := range txtlines {
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:43]: Please combine these two `for` into one. IOW there is no need to have `txtlines`.

File: crit-go/scripts/magic_gen.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	by_val := make(map[uint64]interface{})
+	scanner := bufio.NewScanner(infile)
+	scanner.Split(bufio.ScanLines)
+
+	var txtlines []string
+
+	for scanner.Scan() {
+		txtlines = append(txtlines, scanner.Text())
+	}
+
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:42]: `scanner.Error()` check is missing here.

File: crit-go/scripts/magic_gen.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	for _, eachline := range txtlines {
+		split := strings.Fields(eachline)
+		if len(split) < 3 {
+			continue
+		}
+
+		if itemExists(split, "#define") == false {
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:49]: nit: `if !itemExists(split, "#define")`

File: crit-go/scripts/magic_gen.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+			continue
+		}
+
+		key := split[1]
+		value := split[2]
+		//fmt.Println(reflect.TypeOf(value))
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:55]: debug comment? should be removed

File: crit-go/scripts/magic_gen.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+		if funk.Contains(all_magic, value) == true {
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:56]: nit: remove ` == true`

File: crit-go/scripts/magic_gen.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+			value := all_magic[value]
+			all_magic[key] = value
+		} else {
+			magic[key] = value
+			all_magic[key] = value
+		}
+	}
+
+	for k, v := range magic {
+		//fmt.Println(reflect.TypeOf(v))
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:66]: debug comment? please remove

File: crit-go/scripts/magic_gen.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+		if v == "0x0" || v == "1" || k == "0x0" || v == "1" {
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:67]: Double check for `v == "1" ` does not make sense

File: crit-go/scripts/magic_gen.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+			continue
+		}
+		if strings.Contains(k, "_MAGIC") == true {
+			k = strings.Replace(k, "_MAGIC", "", -1)
+		}
+		vhexedv := hex2int(v)
+		check(err)
+		//fmt.Println(vhexedv)
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:75]: debug print

File: crit-go/scripts/magic_gen.go [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+		by_name[k] = vhexedv
+		by_val[vhexedv] = k
+	}
+	mapper["by_name"] = by_name
+	mapper["by_val"] = by_val
+	mapjsoned, err := json.Marshal(mapper)
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:81]: this is very inefficient. Please use json.NewEncoder() and json.Encode()

File: crit-go/scripts/magic_gen.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+	if s.Index(0).Interface() == item {
+		return true
+	} else {
+		return false
+	}
+}
+
+// function to convert hex string to int
+func hex2int(hexStr string) uint64 {
+	// remove 0x suffix if found in the input string
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:110]: prefix?

File: crit-go/scripts/magic_gen.go [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+	cleaned := strings.Replace(hexStr, "0x", "", -1)
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:111]: strconv.ParseUint() is perfectly capable of parsing a string with `0x` prefix, see https://golang.org/pkg/strconv/#ParseInt

File: crit-go/scripts/magic_gen.go [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+	s := reflect.ValueOf(slice)
+	if s.Kind() != reflect.Slice {
+		panic("Invalid data-type")
+	}
+
+	if s.Index(0).Interface() == item {
+		return true
+	} else {
+		return false
+	}
```
Comment (kolyshkin) [inline crit-go/scripts/magic_gen.go:105]: `return s.Index(0).Interface() == item`

File: crit-go/scripts/magic_gen.go (hunk 15 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go:110]: sorry typo

File: crit-go/scripts/magic_gen.go (hunk 14 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go:81]: slightly inefficient but json.Encode spits out digits,this is more human readable.Although it doesn't really matter to us as have to unmarshal/decode alter anyways.

File: crit-go/scripts/magic_gen.go (hunk 16 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go:111]: tried that,returns 0 stating invalid no. Thus have to clean the prefix first then send to strconv.ParseUint

File: crit-go/scripts/magic_gen.go (hunk 6 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go]: fixed it.

File: crit-go/scripts/magic_gen.go (hunk 7 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go]: error check added

File: crit-go/scripts/magic_gen.go (hunk 8 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go]: done

File: crit-go/scripts/magic_gen.go (hunk 9 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go]: removed all debug comments

File: crit-go/scripts/magic_gen.go (hunk 10 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go]: fixed

File: crit-go/scripts/magic_gen.go (hunk 11 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go]: removed comment

File: crit-go/scripts/magic_gen.go (hunk 12 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go]: removed and changed to single check

File: crit-go/scripts/magic_gen.go (hunk 14 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go]: marshal changed to json.Encode()

File: crit-go/scripts/magic_gen.go (hunk 17 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go]: done

File: crit-go/scripts/magic_gen.go (hunk 5 above)
Comment (Pac23) [reply crit-go/scripts/magic_gen.go]: created a magic-gen directory.

File: crit-go/magic-gen/magic_gen.go [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+		panic("Invalid data-type")
+	}
+
+	return s.Index(0).Interface() == item
+}
+
+// function to convert hex string to int
+func hex2int(hexStr string) uint64 {
+	// remove 0x prefix if found in the input string
+	cleaned := strings.Replace(hexStr, "0x", "", -1)
```
Comment (adrianreber) [inline crit-go/magic-gen/magic_gen.go:101]: is this really necessary? Can't Go convert from 0x123e to integer?

File: crit-go/magic-gen/magic_gen.go [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+	"fmt"
+	"github.com/thoas/go-funk"
+	"os"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+func main() {
+	// checks aregyument correctness
```
Comment (adrianreber) [inline crit-go/magic-gen/magic_gen.go:15]: typo in comment

File: crit-go/magic-gen/magic_gen.go [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+		}
+		if strings.Contains(k, "_MAGIC") == true {
+			k = strings.Replace(k, "_MAGIC", "", -1)
+		}
+		vhexedv := hex2int(v)
+		check(err)
+		by_name[k] = vhexedv
+		by_val[vhexedv] = k
+	}
+	// by_val and by_name ampped to a master map to be converted to json
```
Comment (adrianreber) [inline crit-go/magic-gen/magic_gen.go:73]: typo in comment

**Comment by adrianreber** [comment]
```
$ make
panic: open magic.h: no such file or directory

goroutine 1 [running]:
main.check(...)
	/share/go/src/github.com/checkpoint-restore/go-criu/crit-go/magic-gen/magic_gen.go:83
main.main()
	/share/go/src/github.com/checkpoint-restore/go-criu/crit-go/magic-gen/magic_gen.go:24 +0xba2
exit status 2
make: *** [makefile:21: test/magicgen] Error 1
```

File: crit-go/magic-gen/magic_gen.go (hunk 18 above)
Comment (Pac23) [reply crit-go/magic-gen/magic_gen.go:101]: strconv.ParseInt considers it to be a long string thus ends with "0" or invalid syntax [playground](https://play.golang.org/p/mwtttfzROIW) thus the step of cleaning it first seems to be the only option to get around the problem.

Comment (adrianreber) [reply crit-go/magic-gen/magic_gen.go:101]: What about `n, err := strconv.ParseUint(s, 0, 64)`? That works for me. If you set the base to `0` it works correctly.

File: .travis.yml [hunk 21]
```diff
@@ -23,3 +24,7 @@ script:
   - make build phaul test/test test/phaul test/piggie
   # Run actual test as root as it uses CRIU.
   - sudo make test phaul-test
+  # This builds crit-go
+  - make -C crit-go/magic-gen lint build magicgen test
+  # Run test of crit-go
+  - sudo make test
```
Comment (adrianreber) [inline .travis.yml:30]: This line is not needed, You are already running the test in line 28. Remove the last two lines.

File: crit-go/magic-gen/Makefile [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+		echo FAIL; \
+		exit 1; \
+	fi
+	@rm -f test_input
+	@rm -f test_output
+
+test: test/magicgen
+		
+magicgen: build
+	curl -s https://raw.githubusercontent.com/checkpoint-restore/criu/criu-dev/criu/include/magic.h -o magic.h
```
Comment (adrianreber) [inline crit-go/magic-gen/Makefile:40]: put the download in its own target and make this target dependent on it. This way the file will not be redownloaded for every step:
```
magic.h:
        curl -s https://raw.githubusercontent.com/checkpoint-restore/criu/criu-dev/criu/include/magic.h -o magic.h

magicgen: build magic.h
	./magic-gen magic.h magic.json
	@rm -f magic.h
```

File: crit-go/magic-gen/Makefile [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+
+all: build test magicgen
+
+lint: 
+	@golint magic-gen.go
+
+build:
+	 @$(GO) build -v
+
+test/magicgen:
```
Comment (adrianreber) [inline crit-go/magic-gen/Makefile:20]: Add a dependency on build:
```
test/magicgen: build
```

File: crit-go/magic-gen/Makefile [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+	test_input=`mktemp`
+	test_output=`mktemp`
+	echo "#define TEST_MAGIC 0x12345678" > test_input
+	@rm -f test_output 
+	@$(GO) run magic-gen.go test_input test_output
```
Comment (adrianreber) [inline crit-go/magic-gen/Makefile:25]: Use the binary built by `build`.

./magic-gen test_input test_output

File: crit-go/magic-gen/Makefile [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+
+lint: 
+	@golint magic-gen.go
+
+build:
+	 @$(GO) build -v
+
+test/magicgen:
+	test_input=`mktemp`
+	test_output=`mktemp`
```
Comment (adrianreber) [inline crit-go/magic-gen/Makefile:22]: Remove the two mktemp lines as you are not using the variables below anyway. You are hard-coding the input and output to `test_input` and `test_output`, which is probably okay here.

File: crit-go/magic-gen/Makefile [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+install.tools:
+	if [ ! -x "$(GOBIN)/golint" ]; then \
+		$(GO) get -u golang.org/x/lint/golint; \
+	fi
+
+clean:
+	@rm -f testmagic.json
+	@rm -f test_output
+
+.PHONY: build test clean lint magicgen
```
Comment (adrianreber) [inline crit-go/magic-gen/Makefile:53]: add a newline at the end of the file

File: crit-go/magic-gen/magic-gen.go [hunk 27]
```diff
@@ -0,0 +1,103 @@
...
+	"os"
+	"reflect"
+	"strconv"
+	"strings"
+)
+
+func main() {
+	// checks argument correctness
+	if len(os.Args) != 3 {
+		fmt.Println("Usage: magic_gen.go path/to/magic.h path/to/magic.json")
```
Comment (adrianreber) [inline crit-go/magic-gen/magic-gen.go:17]: I think the resulting binary is called `magic-gen` and not  `magic_gen.go`. Adapt the help output.

**Comment by adrianreber** [comment]
Nice, with integrated travis tests. Looks good. I added a couple more reviews, but then it should be ready to be merged.

**Comment by adrianreber** [comment]
Thanks! Merged! Great work!

**Pull request merged**